gateway-sp-comms/src/single_sp.rs

```diff
@@ -165,64 +166,129 @@ impl SingleSp {
+            //    (we'll never sleep)
+            // 2. Prep is relatively slow (e.g., erasing a flash part)
+            //
+            // We choose a few seconds assuming this polling interval is
+            // primarily hit when the SP is doing something slow.
+            const POLL_UPDATE_STATUS_INTERVAL: Duration =
+                Duration::from_secs(2);
+
+            // Poll SP until update preparation is complete.
+            loop {
+                // Get update status from the SP or give up.
+                let status = match update_status(&inner, component).await {
+                    Ok(status) => status,
+                    Err(err) => {
+                        error!(
+                            log, "update failed: could not get status from SP";
+                            "err" => %err,
+                        );
+                        return;
+                    }
+                };
 
-        // Wait until the SP finishes whatever prep work it needs to do.
-        self.poll_for_update_prepare_status_done(component, stream_id)
-            .await
-            .map_err(UpdateError::Start)?;
-        info!(
-            self.log, "SP update preparation complete";
-            "stream_id" => stream_id,
-        );
+                // Either sleep and retry (if still preparing), break out of our
+                // loop (if prep complete), or fail (anything else).
+                match status {
+                    UpdateStatus::Preparing(sub_status) => {
+                        if sub_status.id == id {
+                            debug!(
+                                log,
+                                "SP still preparing; sleeping for {:?}",
+                                POLL_UPDATE_STATUS_INTERVAL
+                            );
+                            tokio::time::sleep(POLL_UPDATE_STATUS_INTERVAL)
+                                .await;
+                            continue;
+                        }
+                    }
+                    UpdateStatus::InProgress(sub_status) => {
+                        if sub_status.id == id {
+                            info!(
+                                log, "update preparation complete";
+                                "update_id" => %update_id,
+                            );
+                            break;
+                        }
+                    }
+                    UpdateStatus::None
+                    | UpdateStatus::Complete(_)
+                    | UpdateStatus::Aborted(_) => (),
+                }
 
-        let mut image = Cursor::new(image);
-        let mut offset = 0;
-        while !CursorExt::is_empty(&image) {
-            let prior_pos = image.position();
-            debug!(
-                self.log, "sending update chunk";
-                "stream_id" => stream_id,
-                "offset" => offset,
-            );
+                error!(
```

Comment:
Is it really an error if the update was already completed or aborted? Maybe a warning makes more sense?

Reply:
I think I'll make a case of `error!`:

1. In the case of "completed", that means somehow the SP ack'd our request to start an update, but somehow this exact same update was completed before we sent the data to the SP. I can image a sequence that leads to this happening, but it seems awfully unlikely/rare.
2. In the case of "aborted", that means someone aborted an update that we are actively trying to deliver. I think we expect aborting updates to be a thing we do if something has gone wrong, but in this case we think things are still proceeding successfully.

In either case, we thought we were successfully on track to send an update to the SP, but it decided that was no longer the case.

All that said, if you strongly disagree and think this should be `warn!`, I'm not going to die on this hill. :)

Reply:
You've convinced me `error!` is fine.